src/python/zquantum/core/interfaces/functions.py

```diff
@@ -0,0 +1,45 @@
+"""Protocols describing different kinds of functions."""
```

Comment:
I'm not sure whether `cost_function.py` isn't a better place for this? `functions.py` is pretty vague name.
WDYT @simonwa7 ?